In [ ]:
!pip install --upgrade xlrd
from google.colab import drive
drive.mount('/content/gdrive')
import pandas as pd

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 96 kB 2.4 MB/s 
  Attempting uninstall: xlrd
    Found existing installation: xlrd 1.1.0
    Uninstalling xlrd-1.1.0:
      Successfully uninstalled xlrd-1.1.0
Mounted at /content/gdrive


### Carga  el  archivo  que  leíste  en  un  dataFrame.  Elimina  todos  los registros  que  sean  de  Estados Unidos.

In [ ]:
datos = pd.read_excel('/content/gdrive/MyDrive/Colab Notebooks/Modulo 3/DSA03-Material_adicional_file_example_XLSX_5000.xlsx')
datos

Unnamed: 0 First Name   Last Name  Gender        Country  Age  \
0              1      Dulce       Abril  Female  United States   32   
1              2       Mara   Hashimoto  Female  Great Britain   25   
2              3     Philip        Gent    Male         France   36   
3              4   Kathleen      Hanner  Female  United States   25   
4              5    Nereida     Magwood  Female  United States   58   
...          ...        ...         ...     ...            ...  ...   
4995        4996       Roma  Lafollette  Female  United States   34   
4996        4997     Felisa        Cail  Female  United States   28   
4997        4998   Demetria       Abbey  Female  United States   32   
4998        4999     Jeromy        Danz    Male  United States   39   
4999        5000   Rasheeda      Alkire  Female  United States   29   

            Date    Id  
0     15/10/2017  1562  
1     16/08/2016  1582  
2     21/05/2015  2587  
3     15/10/2017  3549  
4     16/08/2016  2468  
...          ...   ...  
4995  15/10/2017  2654  
4996  16/08/2016  6525  
4997  21/05/2015  3265  
4998  15/10/2017  3265  
4999  16/08/2016  6125  

[5000 rows x 8 columns]

In [ ]:
datos2=datos.loc[datos['Country'] != 'United States']
datos2

Unnamed: 0 First Name  Last Name  Gender        Country  Age  \
1              2       Mara  Hashimoto  Female  Great Britain   25   
2              3     Philip       Gent    Male         France   36   
6              7       Etta       Hurn  Female  Great Britain   56   
9             10     Fallon    Winward  Female  Great Britain   28   
10            11    Arcelia     Bouska  Female  Great Britain   39   
...          ...        ...        ...     ...            ...  ...   
4981        4982     Stasia     Becker  Female  Great Britain   34   
4982        4983       Jona    Grindle  Female  Great Britain   26   
4983        4984      Judie   Claywell  Female         France   35   
4986        4987     Kelsie    Wachtel  Female         France   27   
4992        4993      Angel      Sanor    Male         France   24   

            Date    Id  
1     16/08/2016  1582  
2     21/05/2015  2587  
6     15/10/2017  3598  
9     16/08/2016  5486  
10    21/05/2015  1258  
...          ...   ...  
4981  16/08/2016  7521  
4982  21/05/2015  6458  
4983  16/08/2016  7569  
4986  16/08/2016  8642  
4992  15/10/2017  3259  

[2600 rows x 8 columns]

### Muestra  el  número  de  personas  registradas  de  cada  país  (ya  sin  los  de  Estados  Unidos). Se espera una tabla que diga país y número de personas.

In [ ]:
import numpy as np

In [ ]:
datos2['Country'].value_counts()

Great Britain    1400
Francia          1200
Name: Country, dtype: int64

### Actualiza el nombre de los países para que el nombre esté en español.

In [ ]:
datos2.loc[(datos2["Country"]== "France"),'Country'] = 'Francia'
datos2.loc[(datos2["Country"]== "Great Britain"),'Country'] = 'Gran Bretaña'
datos2

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:1817: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


Unnamed: 0 First Name  Last Name  Gender       Country  Age        Date  \
1              2       Mara  Hashimoto  Female  Gran Bretaña   25  16/08/2016   
2              3     Philip       Gent    Male       Francia   36  21/05/2015   
6              7       Etta       Hurn  Female  Gran Bretaña   56  15/10/2017   
9             10     Fallon    Winward  Female  Gran Bretaña   28  16/08/2016   
10            11    Arcelia     Bouska  Female  Gran Bretaña   39  21/05/2015   
...          ...        ...        ...     ...           ...  ...         ...   
4981        4982     Stasia     Becker  Female  Gran Bretaña   34  16/08/2016   
4982        4983       Jona    Grindle  Female  Gran Bretaña   26  21/05/2015   
4983        4984      Judie   Claywell  Female       Francia   35  16/08/2016   
4986        4987     Kelsie    Wachtel  Female       Francia   27  16/08/2016   
4992        4993      Angel      Sanor    Male       Francia   24  15/10/2017   

        Id  
1     1582  
2     2587  
6     3598  
9     5486  
10    1258  
...    ...  
4981  7521  
4982  6458  
4983  7569  
4986  8642  
4992  3259  

[2600 rows x 8 columns]

### Cambia el id de Judie Claywell Female France de 7569 a 99999.


In [ ]:
datos2.loc[(datos2["First Name"]== "Judie")&(datos2["Last Name"]=="Claywell")&(datos2["Gender"]=="Female")&(datos2["Country"]=="Francia"),"Id"] = 99999
datos2

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:1817: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


Unnamed: 0 First Name  Last Name  Gender       Country  Age        Date  \
1              2       Mara  Hashimoto  Female  Gran Bretaña   25  16/08/2016   
2              3     Philip       Gent    Male       Francia   36  21/05/2015   
6              7       Etta       Hurn  Female  Gran Bretaña   56  15/10/2017   
9             10     Fallon    Winward  Female  Gran Bretaña   28  16/08/2016   
10            11    Arcelia     Bouska  Female  Gran Bretaña   39  21/05/2015   
...          ...        ...        ...     ...           ...  ...         ...   
4981        4982     Stasia     Becker  Female  Gran Bretaña   34  16/08/2016   
4982        4983       Jona    Grindle  Female  Gran Bretaña   26  21/05/2015   
4983        4984      Judie   Claywell  Female       Francia   35  16/08/2016   
4986        4987     Kelsie    Wachtel  Female       Francia   27  16/08/2016   
4992        4993      Angel      Sanor    Male       Francia   24  15/10/2017   

         Id  
1      1582  
2      2587  
6      3598  
9      5486  
10     1258  
...     ...  
4981   7521  
4982   6458  
4983  99999  
4986   8642  
4992   3259  

[2600 rows x 8 columns]

### Crea  un  archivo  de  Excel  en  tu  drive,  que  contenga  exactamente  el mismo  primer  renglón  del archivo descargado, pero ahora Inserta 3 registros con gente de México con ids de 5 dígitos que no inicien en cero.

In [ ]:
datos3 = pd.read_excel('/content/gdrive/MyDrive/Colab Notebooks/Modulo 3/Doc2.xlsx')
datos3

Unnamed: 0 First Name Last Name  Gender        Country  Age        Date  \
0           1      Dulce     Abril  Female  United States   32  15/10/2017   

     Id  
0  1562

In [ ]:
from pickle import TRUE
new_row1=pd.Series(["Juan","Frausto","Male","México",25,'14/05/1997',78895])
new_row2=pd.Series(["Max","Arias","Male","México",21,'19/11/1999',45612])
new_row3=pd.Series(["Norma","Avila","Female","México",50,'14/12/1940',12345])
new_rows_df = pd.DataFrame(new_row1)
new_rows_df = pd.concat([new_rows_df, new_row2], axis=1)
new_rows_df = pd.concat([new_rows_df, new_row3], axis=1)
new_rows_df = new_rows_df.T
new_rows_df.columns = ["First Name",'Last Name', 'Gender', 'Country', 'Age', 'Date', 'Id']
#new_rows_df.reset_index(drop = TRUE)
new_rows_df

First Name Last Name  Gender Country Age        Date     Id
0       Juan   Frausto    Male  México  25  14/05/1997  78895
0        Max     Arias    Male  México  21  19/11/1999  45612
0      Norma     Avila  Female  México  50  14/12/1940  12345

### Carga este segundo archivo como un dataFrame y concaténalo a los datos del archivo descargado.

In [ ]:
datos4=pd.concat([datos2, new_rows_df], ignore_index = True)
datos4

Unnamed: 0 First Name  Last Name  Gender       Country Age        Date  \
0            2.0       Mara  Hashimoto  Female  Gran Bretaña  25  16/08/2016   
1            3.0     Philip       Gent    Male       Francia  36  21/05/2015   
2            7.0       Etta       Hurn  Female  Gran Bretaña  56  15/10/2017   
3           10.0     Fallon    Winward  Female  Gran Bretaña  28  16/08/2016   
4           11.0    Arcelia     Bouska  Female  Gran Bretaña  39  21/05/2015   
...          ...        ...        ...     ...           ...  ..         ...   
2598      4987.0     Kelsie    Wachtel  Female       Francia  27  16/08/2016   
2599      4993.0      Angel      Sanor    Male       Francia  24  15/10/2017   
2600         NaN       Juan    Frausto    Male        México  25  14/05/1997   
2601         NaN        Max      Arias    Male        México  21  19/11/1999   
2602         NaN      Norma      Avila  Female        México  50  14/12/1940   

         Id  
0      1582  
1      2587  
2      3598  
3      5486  
4      1258  
...     ...  
2598   8642  
2599   3259  
2600  78895  
2601  45612  
2602  12345  

[2603 rows x 8 columns]

### Vuelve a mostrar el número de personas registradas de cada país.

In [ ]:
datos4['Country'].value_counts()

Gran Bretaña    1400
Francia         1200
México             3
Name: Country, dtype: int64

In [58]:
!pip install xlsxwriter

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 149 kB 5.2 MB/s 


In [65]:
datos4.to_excel("/content/gdrive/MyDrive/Colab Notebooks/Modulo 3/Actívate_02.ipynb", sheet_name="gentePorPais", engine="xlsxwriter")

ValueError: ignored